### ISSUE TITLE
Unexpected behavior of self.add_instance when arraying and with orientation not R0

### ISSUE THREAD

**Issue by sbuchbinder** [ucb-art/BAG_framework#33]
What is the expected behavior of calling TemplateBase.add_instance on an array (ie, nx>1 or ny>1), when there is also a rotation?

Is each element rotated in place, or is the whole array treated as a mosaic, and rotated?

Either way, the behavior does not currently seem correct?

Below, I place an instance of a master, which I array in a row of 5 columns.  The only difference is that I orient the second case by R90 (and place at a different location).

```python
        self.add_instance(
            master=master,
            loc=(0, 0),
            orient='R0',
            nx=5,
            spx=10,
            unit_mode=False
        )

        self.add_instance(
            master=master,
            loc=(0, 20),
            orient='R90',
            nx=5,
            spx=10,
            unit_mode=False
        )
```

As can be seen, it looks like not only do the sub-instances get rotated, but also translated?
![image](https://user-images.githubusercontent.com/18271765/50255676-b275d800-03a7-11e9-81ac-56d9912b6c27.png)

I would expect either:
Treat array as mosaic and rotate:
![image](https://user-images.githubusercontent.com/18271765/50255864-7b53f680-03a8-11e9-8ea7-f89f9d17704d.png)

or
Rotate each sub-instance in place:
![image](https://user-images.githubusercontent.com/18271765/50256170-e520d000-03a9-11e9-91eb-9b5961f310c6.png)

**Comment by pkerichang** [comment]
The bag behavior is rotate first, then array.  However, since cadence does
it the other way, I need to hack the values when passing to cadence.  Seems
like my hack doesn't work for 90 rotations.

Thanks for catching this bug.

Eric

**Comment by pkerichang** [comment]
Yes, I believe so.  Have you fix this yet, or should I take a look?

**Comment by sbuchbinder** [comment]
I have not looked into it further.

**Comment by sbuchbinder** [comment]
@pkerichang --
This coincidentally became relevant to BPG development this week, so I wanted to bring it up again.

In addition to behavior of rotating instances with arrays (which I think might be fixable in the InstanceInfo?), can we also discuss the behavior of transforming Arrayable objects?  

I believe that 90 degree rotations are not currently supported, and that even mirroring / R180 transformations are not correct. 

As an example:
The correct desired behavior (if I change the blocks to not use any arraying of rects):
![image](https://user-images.githubusercontent.com/18271765/53284197-ab4fab00-3705-11e9-8e0b-32ad82d1335b.png)

The current behavior when using arraying:
![image](https://user-images.githubusercontent.com/18271765/53284208-e356ee00-3705-11e9-9d1c-8e1fcd036e1b.png)


The objects are getting placed by calling the transform method of the Rect class. But if the rect is arrayed, transform does not adjust the nx, ny, spx, spy.


I think that the problem can be fixed with code similar to below:


```
def transform_arrays(nx: int,
                     ny: int,
                     spx: float,
                     spy: float,
                     orient: str,
                     ) -> Tuple[int, int, float, float]:
    """
    Transforms the array parameters of an object based on the passed transformation rotation.

    Parameters
    ----------
    nx : int
        number of arrayed rectangles in the x-direction
    ny : int
        number of arrayed rectangles in the y-direction
    spx : float
        space between arrayed rectangles in the x-direction
    spy : float
        space between arrayed rectangles in the y-direction
    orient : str
        the transformation orientation string

    Returns
    -------
    (nx, ny, spx, spy) : Tuple[int, int, float, float]
        The transformed array parameters of the object
    """

    mat = transform_table[orient]
    sp_new = np.dot(mat, np.array([spx, spy])).astype(float)
    n_new = np.dot(mat, np.array([nx, ny])).astype(int)

    return abs(n_new.item(0)), abs(n_new.item(1)), sp_new.item(0), sp_new.item(1)
```

However, this has its own problems, because BBoxArray does not support negative spx, spy, and also does not support 90deg rotations.

Do you have any thoughts on how to support transformation of Arrayables?
